crates/engine/tree/src/tree.rs

```diff
@@ -5,19 +5,34 @@ use reth_blockchain_tree_api::{error::InsertBlockError, InsertPayloadOk};
 use reth_engine_primitives::EngineTypes;
 use reth_errors::ProviderResult;
 use reth_payload_validator::ExecutionPayloadValidator;
-use reth_primitives::{Block, SealedBlock, SealedBlockWithSenders, B256};
+use reth_primitives::{Block, BlockNumber, SealedBlock, SealedBlockWithSenders, B256};
 use reth_provider::BlockReader;
 use reth_rpc_types::{
     engine::{CancunPayloadFields, ForkchoiceState, PayloadStatus, PayloadStatusEnum},
     ExecutionPayload,
 };
-use std::{marker::PhantomData, sync::Arc};
+use std::{
+    collections::{BTreeMap, HashMap},
+    marker::PhantomData,
+    sync::Arc,
+};
 use tracing::*;
 
+/// Represents an executed block stored in-memory.
+#[derive(Clone, Debug)]
+struct ExecutedBlock {
+    block: Arc<SealedBlockWithSenders>,
+    state: Arc<()>,
+    trie: Arc<()>,
+}
```

Comment:
any reason why the struct is not just Arc'ed in data structures that use it, since all the fields are Arc

Reply:
we will need to clone different states for different purposes. e.g. sealed block should be returned for `block_by_hash` request